### ISSUE TITLE
Warn when memset size/value args may be swapped

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#14179]
|  |  |
| --- | --- |
| Bugzilla Link | [13807](https://llvm.org/bz13807) |
| Version | trunk |
| OS | Linux |
| CC | @DougGregor,@gribozavr |

## Extended Description 
Clang doesn't warn when the user inverts the 2nd and 3rd argument to memset:

int i;
memset(&i, sizeof(i), 0); // user intends memset(&i, 0, sizeof(i))

If the user wrote the literal 0 as the memset length that's clearly wrong (assuming no macros got in the way - ie: the source location "FileID" of both the literal 0 and the word 'memset' are the same (if I understand SourceLocations correctly that statement should make sense)) & they probably meant to invert the arguments. We could potentially use the value of the second parameter as a further hint, but I'm not sure how/when it would be a good signal (honestly I expect the vast majority of memsets to be setting to zero, but I could be wrong -perhaps someone does have a need to set the value of memory to the size of some time using memset...)

[I should probably look at llvm/llvm-project#10349  for some code related to memset warnings as a starting point to adding this warning]

**Comment by wheatman** [comment]
This was fixed between 6 and 7
https://godbolt.org/z/EsnseYje4

code
```
#include <string.h>
int main() {
    int i;
    memset(&i, sizeof(i), 0); // user intends memset(&i, 0, sizeof(i))
}
```
new warning
```
<source>:4:23: warning: 'size' argument to memset is '0'; did you mean to transpose the last two arguments? [-Wmemset-transposed-args]
memset(&i, sizeof(i), 0); // user intends memset(&i, 0, sizeof(i))
                      ^
<source>:4:23: note: parenthesize the third argument to silence
1 warning generated.
Compiler returned: 0
```
